### DISCUSSION TITLE
Replace disk, same server

### DISCUSSION THREAD

**Discussion by guestisp** [moosefs/moosefs#587] [Q&A]
Is possibile to mark a disk for removal and then start migrating data to another disk on the same chunkserver?

It would be much better and faster to sync 2 disks locally without involving any network operation

**Comment by oszafraniec** [comment]
Mayby test this... "_'<' means that all data from this hard drive should be moved to other hard drives_"

$ cat /etc/mfs/mfshdd.cfg
```
 # This file keeps definitions of mounting points (paths) of hard drives to use with chunk server.
 # A path may begin with extra characters which swiches additional options:
 #  - '*' means that this hard drive is 'marked for removal' and all data will be replicated to other hard  drives (usually on other chunkservers)
 #  - '<' means that all data from this hard drive should be moved to other hard drives
 #  - '>' means that all data from other hard drives should be moved to this hard drive 
 #  - '~' means that significant change of total blocks count will not mark this drive as damaged
 # If there are both '<' and '>' drives then data will be moved only between these drives
```

**Comment by guestisp** [comment]
`<` is clear and it's what I need.

But `>` is not clear to me. How many (and which) disks are you replicating to the new one ? Moving 2 or more disks to a single one, will reduce fault tollerance. If data1 is stored on disk1 and data2 is stored on disks2, in case of disk1 failure, you'll only loose data1. If you move both to disk3, when disk3 fails, you'll loose both data1 and data2.

Am I miss something ?

**Comment by guestisp** [comment]
> ` # If there are both '<' and '>' drives then data will be moved only between these drives`

So:

```
/mnt/hdd1
< /mnt/hdd2
> /mnt/hdd3
```

Will move only data from hdd2 to hdd3 ?

**Comment by oszafraniec** [comment]
i dont know... just try :)

I dont know how many CHSRV you have, but letting MFS to do it via network, event 1G, is fast enough. 4TB drive needs just ~2h to replicate on our system.

**Comment by guestisp** [comment]
My test cluster will have a 3 chunkservers with goal set to 3.
One thing i've noticed when I've tested LizardFS is that goal 3 on a 3 chunkservers cluster is not a good idea, as there isn't room for movements.

In example, in Lizard, you won't be able to mark a disk for removal even if on the same chunkserver there are brand new and available disks, because lizard will never move the same data on the same chunkserver (thus, removing from disk0 and writing to disk1 on the same server is impossibile)

MooseFS seems to be able to do this and for me is a **huge** plus.

**Comment by pkonopelko** [comment]
> So:
> 
> ```
> /mnt/hdd1
> < /mnt/hdd2
> > /mnt/hdd3
> ```
> Will move only data from hdd2 to hdd3 ?

Yes

**Comment by guestisp** [comment]
Another question: what happens if MooseFS mark a disk as damaged due to too many I/O errors and a new disk is available on the same chunkserver ? The removal will migrate locally or always via network ?

**Comment by acid-maker** [comment]
Removal will migrate always via network - I see at least two reasons:

1. This is much faster because all servers/disks are involved. Usually damaged 1TB disk is replicated in about 15 minutes - using only local transfer it usually takes about day (from damaged disk probably much longer).

2. When disk goes RO there is usually reason for that and the last thing we want is to use this disk more than needed.

**Comment by guestisp** [comment]
Not all removals are done due to a disk failure.

Yes, network is faster if you replicate a removed drive reading from multiple sources and writing to multiple disks, but in my case, if you have a 3 nodes cluster with goal=3, you can't move to a different node, because you don't have a spare node available to use as destination

You are forced to replicate to a different disk on the same node. Is this possible? If yes, network replication doesn't provide any benefits because you are still limited by a single disk write speed, so, replicating over network something that could be done faster locally only add more stress to the network layer and the entire cluster

**Comment by acid-maker** [comment]
As somebody mentioned it before - you can do it manually using '>' and '<' - this is why we added it. If you want to migrate data on purpouse then you can. In case of emergency it is usually bad idea.

Maybe for specific case (goal = number of CS) it would work, but this is not recommended MFS installation. Doing this would mean a lot of code ONLY for this setting. I'd rather spend this time to implement more interesting features :)

**Comment by 4Dolio** [comment]
I was not aware of the < > and other flags until this thread. Very handy feature for planned disk upgrades, where I have been using my own local rebalancing scripts, the cheeky little scripts will be missed. ;)

**Comment by guestisp** [comment]
@acid-maker I totally agree with you. What I'm asking is: what happens when MooseFS detect a disk failure and mark that for removal ? (i've seen in the code that Moose is able to mark a disk for removal automatically in case I/O errors are greater than an X value in Y seconds).

For planned disk upgrades, `>` and `<` would be great and it's exactly what I need.

**Comment by acid-maker** [comment]
There are two different statuses: "Mark for removal" and "Damaged".

"Mark For Removal" status is set manually and it means "I want to remove this drive ASAP" - MFS in such case makes enough copies of chunks from this disk on all other CS'es that after removing such drive all goals should be fulfilled.

"Damaged" status is set automatically whenever more than X I/O errors happens in Y seconds - in such case CS just report to MFS that all chunks from damaged drive are missing.

As a result master in both cases starts replication, but in first case disk is still used in normal read/write operations (only new chunks are not created on "MFR" disk), in second case disk is not used at all - it is left in data structures only for CLI/CGI reports.
In the first case disk can be used as standard replication source (because MFS knows about all chunks on this drive), in the second case disk can't be used as replication source (because MFS "thinks" that there are no chunks there).

**Comment by guestisp** [comment]
Clear
But what if I have to mark for removal a disk with issues on filesystem that was placed in readonly? Any write operation will generate tons of errors...

Why not avoiding to write on a marked for removal disk? If i want to remove it, why writes should be allowed?

Another issue is: i know that a disk can fail shortly (smart errors and so on). I want to replace it ASAP, thus i mark it for removal
Writes should be denied ad they add a useless load on an "almost-failed" disk.

Similiar to an emergency backup: writes are the last thing you would want on a failed disk

**Comment by acid-maker** [comment]
First of all broken disk in MFR mode won't produce tons of errors. New chunk will not be created there, so number of writes is reduced only to modification of old chunks (this is quite rare operation in FS).

Why we still want to modify old chunks? Because otherwise we would need to increase version on all other copies and then delete this "old-versioned" copy from MFR disk (which is also kind of write operation).

What you really want is probably "READ ONLY" mode. This should be something new - we shouldn't modify MFR behaviour. I think that this RO mode should be dedicated especially for broken disks.

And yes it is on our road map - admins of our own MFS realy want this option for a long time :)

**Comment by 4Dolio** [comment]
Interesting, read only mode... Maybe related, not really, could follower masters perhaps be used for read only clients?  I'll come up with a reason later... ;)

**Comment by guestisp** [comment]
So, is MooseFS a COW filesystem ? In example, a VM image hosted on MooseFS, will always produce new chunks or also change the existing ones?

**Comment by 4Dolio** [comment]
I think it changes the existing ones, for any given 'location' within the file which previously already had data. Or it creates new chunks as needed as a file is written to new regions previously devoid of data.

Or maybe you are referring to snapshots? Those are COW for sure, I suspect individual chunk are also each COW in nature.. so if you have a snapshot the original Chunks becomes COW I guess?

**Comment by guestisp** [comment]
> I think it changes the existing ones, for any given 'location' within the file which previously already had data. Or it creates new chunks as needed as a file is written to new regions previously devoid of data.

So, changing an existing chunk is not so rare, particularly with VM hosting. So my request is legit. If you mark a disk for removal, you should not write to it at all.

> Or maybe you are referring to snapshots?

No, i'm referring to normal chunks, not shapshots.

**Comment by acid-maker** [comment]
@guestisp It depends. As **4Dolio** said in normal files MFS just modifies chunks, but when you make snapshot of a file then during any modification first time MFS makes copy of shared chunks and then modifies it (COW).

@4Dolio - We were thinking of using follower as a read-only master. Main operations in MFS are **lookup** and **getattr** - both do not modify metadata, so can be performed by follower. The only problem is that follower is little behind the leader, so it could lead to synchronization issues. We can of course introduce new flag to directories 'allow answers from follower' (or something like that), but In my opinion nobody will use that. Does anybody ever used any flag using `mfsseteattr`?

@guestisp I wrote that modification are rare - on normal file system when you use files as files they are rare - we usually create files, write them and then only read them, but of course in case of VM environement you probably have mainly reads and modifications of existing chunks. You are abosolutely right.

BTW. I'm thinking that maybe such RO mode should replace current "damaged" mode? Instead of informing master that chunks are missing maybe CS should inform master that chunks are on "damaged" disk. In such case master should replicate this chunk using other replicas and totally avoid I/O on such disk (do not create chunks, do not delete them, do not modify them and read only as last resort).

**Comment by guestisp** [comment]
> We were thinking of using follower as a read-only master. Main operations in MFS are lookup and getattr - both do not modify metadata, so can be performed by follower. The only problem is that follower is little behind the leader, so it could lead to synchronization issues.

IMHO, making master multithreaded would be enough. With a multithreaded master you'll be able to use multiple CPUs (and consequently twice the RAM for metadata). Keeping everything in RAM, on a multithreaded process should be enough to properly manage hundred thousands (or even millions) of requests per seconds. You'll have to saturate two CPUs with small requests.

Even better, allowing master to lookup from disks (SSDs or NVMe) for less accessed metadatas will preserve tons of ram without affecting performance. (a sort of "internal swap")

**Comment by guestisp** [comment]
> I'm thinking that maybe such RO mode should replace current "damaged" mode? Instead of informing master that chunks are missing maybe CS should inform master that chunks are on "damaged" disk. In such case master should replicate this chunk using other replicas and totally avoid I/O on such disk (do not create chunks, do not delete them, do not modify them and read only as last resort).

Looks very good to me. particularly the last part: "read only as last resort". In this case, if replication could be made totally from network, the bad disks is not accessed at all (preserving it as much as possible). If a chunk replication can't be made from network, you could try to read from the bad disk. If everything goes bad, sysop can try to migrate manually with rsync

**Comment by acid-maker** [comment]
@guestisp "threads are evil"

There are other ways to improve master efficiency. Threads are main source of most bugs in modern software - we do not want that - especially in mfsmaster.

Using NVMe instead of RAM in mfsmaster is of course another story. Maybe in the future.

**Comment by guestisp** [comment]
But, what about allowing master to swap locally less-accessed data ? Obviously, it's a sysop decision to only use RAM or even local disks (with slow performance).

What happens, currently, if master start to swap ? Only a performance drop or any other issue?

**Comment by 4Dolio** [comment]
I am not sure, but i believe when a master dumps metadata to disk, hourly (or tuned).  When it dumps it clones the main process in RAM, doubling the used memory with the second process, and dumps from the copy.

But if there is not enough RAM to do this then the main master process goes into a blocking mode while it dumps.  I never give my masters enough swap space to have that as an option, maybe I should try it out on a NVMe backed master..

**Comment by acid-maker** [comment]
@guestisp We have tested it. We even use NVMe only for swap and results were poor. Performance of mfsmaster dropped down significantly.

@4Dolio fork during saving metadata doesn't double used memory because after fork whole memory is shared between processes and only changed pages are duplicated (COW). Usually it increases needed memory of about 10% (depends on number of operations and speed of metadata save). Remember that on Linux you should do:
`echo 1 > /proc/sys/vm/overcommit_memory`
to switch off algorithm that prevents forking of huge (RAM consuming) processes

**Comment by pkonopelko** [comment]
@4Dolio: you can also read a bit more about Master forking and `overcommit_memory` in Best Practices [here](https://moosefs.com/support/#best-practices) - it is Best Practice no. 7 "`overcommit_memory` on Master Servers (Linux only)".

Best,
Peter / MooseFS Team
